fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -155,10 +155,30 @@ class FullyShardedDataParallel(nn.Module):
             *``cpu_offload``*.
         bucket_cap_mb (int, Optional):
             FSDP will bucket parameters so that gradient reduction can
-            potentially overlap with backward computation. bucket_cap_mb
-            controls the bucket size in MegaBytes (MB). Buckets are sub-divided
-            based on world_size, so the max shard size is roughly
-            ``bucket_cap_mb / world_size``. Values <= 0 disable bucketing.
+            be more efficient for small tensors.
+            ``bucket_cap_mb`` controls the bucket size in MegaBytes (MB). Buckets
+            are sub-divided based on world_size, so the max shard size is roughly
+            ``bucket_cap_mb / world_size``. When nested FSDP is used,
+            each FSDP instance will have a separate set of buckets (1 per bucket
+            per tensor <type, device, process group> tuple). Large gradient tensors
```

Comment:
The setup/teardown functionality looks good.

But I thought we share the ReduceScatterBucketer instance across all nested FSDP instances (similar to how we share streams), so I thought there shouldn't be separate buckets for separate FSDP instances: https://github.com/facebookresearch/fairscale/blob/4dce2d62df4b8aeee35ff156cb83bfa6e6d1c970/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L938-L940

Reply:
Thanks for catching this! My comments are certainly incorrect.

This is very subtle. I commented out the the teardown and the memory for the unit test became:
```
{
    "iter 0: start": 3,
    "iter 0: after fwd": 340,
    "iter 0: after loss": 340,
    "iter 0: after bwd": 258,
    "iter 0: after step": 258,
    "iter 0: done": 196,
    "iter 1: start": 196,
    "iter 1: after fwd": 533,
    "iter 1: after loss": 533,
    "iter 1: after bwd": 258,
    "iter 1: after step": 258,
    "iter 1: done": 196,
    "iter 2: start": 196,
    "iter 2: after fwd": 533,
    "iter 2: after loss": 533,
    "iter 2: after bwd": 258,
    "iter 2: after step": 258,
    "iter 2: done": 196,
}
```
vs. with `teardown`
```
{
    "iter 0: start": 3,
    "iter 0: after fwd": 340,
    "iter 0: after loss": 340,
    "iter 0: after bwd": 66,
    "iter 0: after step": 66,
    "iter 0: done": 3,
    "iter 1: start": 3,
    "iter 1: after fwd": 340,
    "iter 1: after loss": 340,
    "iter 1: after bwd": 66,
    "iter 1: after step": 66,
    "iter 1: done": 3,
    "iter 2: start": 3,
    "iter 2: after fwd": 340,
    "iter 2: after loss": 340,
    "iter 2: after bwd": 66,
    "iter 2: after step": 66,
    "iter 2: done": 3,
}
```

It increased by 193MB, which close to 5 * (25 + 12.5) = 187.5. The `5` is actually number of different process groups instances used by the model.

The test is already using a global PG. But `auto_wrap_bn` was not using the right global PG.

It seems we should index the buckets dict using a key that represent the process group with its members in this line:

https://github.com/facebookresearch/fairscale/blob/9b79cc020981a9b6e0b13570ede36a080a69c7b5/fairscale/utils/reduce_scatter_bucketer.py#L144

But I can't seem to find an API of the process group that can turn it into its members. Ideas? Maybe leave it as is?

Reply:
I updated the comment but hasn't change the `key = ...` line.

Reply:
I will merge this as it with a TODO here once the CI passes so that vissl can use this.